### PR TITLE
heap.Client emits an event when an error occurs.

### PR DESCRIPTION
**Pull request by pwnall**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -32,9 +32,41 @@ Create an API client.
...
+to do more work when the call completes.
+
+```javascript
+heap.addUserProperties('user-identity', { plan: 'premium1' })
+.then(function() {
+  // Do more work.
+});
+```
+
+Set up an [event](https://nodejs.org/api/events.html) listener to log Heap API
```
Comment (macobo) [inline README.md:58]: will this also fire for validations?

File: lib/client.js [hunk 2]
```diff
@@ -140,12 +166,20 @@ Client.prototype._request = function(options, callback) {
...
+  var _this = this;
   return new Promise(function(resolve, reject) {
     request(options, function (error, response) {
-      if (!error && response.statusCode >= 400)
-        error = Error("Heap API server error " + response.statusCode);
+      if (error) {
+        response = null;
+      } else {
+        if (response.statusCode >= 400) {
+          error = Error("Heap API server error " + response.statusCode + ": " +
```
Comment (macobo) [inline lib/client.js]: perhaps attach the response object to the error? I think node-pg does something like this for some pg errors for an awesome effect.

File: README.md (hunk 1 above)
Comment (pwnall) [inline README.md:58]: Nope, validations throw exceptions immediately, and the caller does have to catch them.

File: lib/client.js (hunk 2 above)
Comment (pwnall) [inline lib/client.js]: Should I throw a custom error type with a stable API here?

File: README.md (hunk 1 above)
Comment (pwnall) [inline README.md:58]: Resolved: now validation errors are handled like all other error types.

File: lib/client.js (hunk 2 above)
Comment (pwnall) [inline lib/client.js]: Resolved: the error event data is now a custom exception that includes the request and the client reference.

**Pull request merged**